detectors/gcp/cloud-run.go

```diff
@@ -0,0 +1,124 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"cloud.google.com/go/compute/metadata"
+
+	"go.opentelemetry.io/otel/label"
+	"go.opentelemetry.io/otel/sdk/resource"
+	"go.opentelemetry.io/otel/semconv"
+)
+
+const serviceNamespace = "cloud-run-managed"
+
+// The minimal list of metadata.Client methods we use. Use an interface so we
+// can replace it with a fake implementation in the unit test.
+type metadataClient interface {
+	ProjectID() (string, error)
+	Get(string) (string, error)
+	InstanceID() (string, error)
+}
+
+// CloudRun collects resource information of Cloud Run instance.
+type CloudRun struct {
+	mc     metadataClient
+	onGCE  func() bool
+	getenv func(string) string
+}
+
+// compile time assertion that CloudRun implements the resource.Detector
+// interface.
+var _ resource.Detector = (*CloudRun)(nil)
+
+// NewCloudRun creates a CloudRun detector.
+func NewCloudRun() *CloudRun {
+	return &CloudRun{
+		mc:     metadata.NewClient(nil),
+		onGCE:  metadata.OnGCE,
+		getenv: os.Getenv,
+	}
+}
+
+// for test only
+func (c *CloudRun) setupForTest(mc metadataClient, ongce func() bool, getenv func(string) string) {
```

Comment:
This probably belongs in the _test.go file. In addition to making the detector more readable, it would ensure test functions are not used in production code.

Reply:
Moved to _test.go and made it a function instead of a method.